kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java

```diff
@@ -0,0 +1,285 @@
+  }
+
+  private static RemoteSessionTable<?> createMongo(
+      final ResponsiveSessionParams params,
+      final SessionClients clients,
+      final SessionSegmentPartitioner partitioner
+  ) throws InterruptedException, TimeoutException {
+    final ResponsiveMongoClient client = clients.mongoClient();
+
+    switch (params.schemaType()) {
+      case SESSION:
+        return client.sessionTable(params.name().tableName(), partitioner);
+      default:
+        throw new IllegalArgumentException(params.schemaType().name());
+    }
+  }
+
+  private static RemoteSessionTable<?> createRemoteSessionTable(
+      final ResponsiveSessionParams params,
+      final SessionClients sessionClients,
+      final SessionSegmentPartitioner partitioner
+  ) throws InterruptedException, TimeoutException {
+    switch (sessionClients.storageBackend()) {
+      case CASSANDRA:
+        return createCassandra(params, sessionClients, partitioner);
+      case MONGO_DB:
+        return createMongo(params, sessionClients, partitioner);
+      default:
+        throw new IllegalStateException("Unexpected value: " + sessionClients.storageBackend());
+    }
+  }
+
+  @SuppressWarnings("rawtypes")
+  public SessionOperationsImpl(
+      final InternalProcessorContext context,
+      final ResponsiveSessionParams params,
+      final RemoteSessionTable table,
+      final CommitBuffer<SessionKey, ?> buffer,
+      final TopicPartition changelog,
+      final ResponsiveStoreRegistry storeRegistry,
+      final ResponsiveStoreRegistration registration,
+      final ResponsiveRestoreListener restoreListener,
+      final long initialStreamTime
+  ) {
+    this.context = context;
+    this.params = params;
+    this.table = table;
+    this.buffer = buffer;
+    this.changelog = changelog;
+    this.storeRegistry = storeRegistry;
+    this.registration = registration;
+    this.restoreListener = restoreListener;
+    this.initialStreamTime = initialStreamTime;
+  }
+
+  @Override
+  public long initialStreamTime() {
+    return initialStreamTime;
+  }
+
+  @Override
+  public void put(
+      SessionKey key,
+      final byte[] value
+  ) {
+    this.buffer.put(key, value, this.context.timestamp());
+  }
+
+  @Override
+  public void delete(final SessionKey key) {
+    this.buffer.tombstone(key, this.context.timestamp());
+  }
+
+  @Override
+  public byte[] fetch(final SessionKey key) {
+    final Result<SessionKey> localResult = this.buffer.get(key);
+    if (localResult != null) {
+      return localResult.isTombstone ? null : localResult.value;
+    }
+
+    final var remoteResult = this.table.fetch(
+        this.changelog.partition(),
+        key.key,
+        key.sessionStartMs,
+        key.sessionEndMs
+    );
+    return remoteResult;
+  }
+
+  @Override
+  public KeyValueIterator<Windowed<Bytes>, byte[]> fetchAll(
+      final Bytes key,
+      final long earliestSessionEnd,
+      final long latestSessionEnd
```

Comment:
parameter name should be latestSessionStart

Reply:
Same as for the RemoteSessionTable interface, I only filter based on session end, so I pass in [earliest, latest] as a way to lower + upper bound the session end.
The latestSessionStart timestamp is only used as a last-minute filter on the results from filtering the session end.